get.js

```diff
@@ -166,19 +156,24 @@ function getStream (cache, key, opts) {
         ev === 'integrity' && cb(entry.integrity)
         ev === 'size' && cb(entry.size)
       })
-      pipe(
-        read.readStream(
-          cache,
-          entry.integrity,
-          opts.concat({
-            size: opts.size == null ? entry.size : opts.size
-          })
-        ),
-        memoStream,
-        stream
+
+      const src = read.readStream(
+        cache,
+        entry.integrity,
+        opts.concat({
+          size: opts.size == null ? entry.size : opts.size
+        })
       )
+
+      if (opts.memoize) {
+        const memoStream = new Collect.PassThrough()
+        memoStream.on('collect', data => memo.put(cache, entry, data, opts))
+        stream.unshift(memoStream)
+      }
+      stream.unshift(src)
```

Comment:
`minipass-pipeline` streams manage a list of streams that are all piped together, and present a single Minipass stream interface on top.  So, writing goes to the `head` of the pipeline, and reading comes from the `tail`.  If the `tail` is not a readable (ie, duplex or PassThrough) stream, then the Pipeline just shoves data into it and that's that.  If the `head` isn't a writable (duplex) then you can't write to the pipeline.

`Pipeline.push(stream)` adds it to the end of the pipeline.  `Pipeline.unshift(stream)` ads it to the start.

So, these three are all equivalent:

```js
// provide to constructor
p = new Pipeline(reader, transform, writer)

// using push
p = new Pipeline()
p.push(reader) // {reader}
p.push(transform) // {reader, transform}
p.push(writer) // {reader, transform, writer}

// using unshift
p = new Pipeline()
p.unshift(writer) // {writer}
p.unshift(transform) // {transform, writer}
p.unshift(reader) // {reader, transform, writer}

// using all three things!
p = new Pipeline(transform) // {transform}
p.unshift(reader) // {reader, transform}
p.push(writer) // {reader, transform, writer}
```

In other words, a Pipeline is pushing _streams_ into the set, using push/unshift like an Array would, to add to the ordered list.

Node-core Readable streams have a "push" method which is a very different thing.  That method is for pushing _data_ into the stream itself.  And, like the push/unshift paradigm from arrays, `s.push(data)` adds the data into the end of the queue, but `s.unshift(data)` adds the data to the _start_ of the queue, which leads to some very confusing and weird scenarios that are really complicated to manage.

Reply:
Ok, well... I wrote a whole explainer here, and then GitHub dropped it, so you get the short version:

`stream` here is a [minipass-pipeline](http://npm.im/minipass-pipeline).  `push` adds a stream to the end of the pipeline.  `unshift` adds a stream to the start of the pipeline.

This is different from node-core's `stream.push(chunk)` which queues a chunk of _data_ top be emitted.  In Minipass lingo, you'd use `this.write(data)`, since they're passthrough streams (or `super.write(data)` if the `write` method has been overridden, which is pretty common.)

Reply:
ok, now the original thing I wrote is back.

Internet gremlins.

Reply:
oooh I get it, just mixed up the minipass api with node's Readable/Writable 😅 

I'm really lucky I got the long version, very instructive, thank you so much for it ❤️